File: Medium/string_to_integer_atoi.js
/**
 * @param {string} s
 * @return {number}
 */
var myAtoi = function (s) {
    s = s.trim();

    if(s.length == 0 || s.length > 200) {
        return 0;
    }

    let collectString = "";
    let isNegative = s[0] == "-";

    for(let i = 0; i < s.length; i++) {
        if ((i == 0 && isNegative) || (i == 0 && s[i] == '+')) {
            continue;
        }

        if(s[i].match(/^[0-9]/g) == null) {
            break;
        }

        collectString += s[i];
    }

    let resultInt = parseInt(collectString);
    if(!resultInt) {
        return 0;
    }

    if(isNegative) {
        resultInt = -(collectString)
    }   

    if(resultInt < -(2**31)) {
		return -(2**31);
	} else if (resultInt > (2**31)-1) {
		return (2**31) - 1;
	}
    
    return resultInt
};

console.log(myAtoi("1337c0d3"));
console.log(myAtoi("words and 987"));
console.log(myAtoi("  -042"));
console.log(myAtoi("20000000000000000000"));
console.log(myAtoi("   +0 123"));
console.log(myAtoi("21474836460"));

File: Medium/string_to_integer_atoi.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func myAtoi(s string) int {
	s = strings.TrimSpace(s)

	if len(s) == 0 || len(s) > 200 {
		return 0
	}

	var collectString string
	isNegative := string(s[0]) == "-"

	for i := 0; i < len(s); i++ {
		if (i == 0 && isNegative) || (i == 0 && string(s[i]) == "+") || string(s[i]) == "" {
			continue
		}

		_, err := strconv.Atoi(string(s[i]))
		if err != nil {
			break
		}

		collectString = collectString + string(s[i])
	}

	resultInt, err := strconv.Atoi(collectString)
	if err != nil {
		if len(collectString) > 5 {
			if isNegative {
				return -(1 << 31)
			} else {
				return (1 << 31) - 1
			}
		}
		return 0
	}

	if isNegative {
		resultInt = -(resultInt)
	}

	if resultInt < -(1 << 31) {
		return -(1 << 31)
	} else if resultInt > (1<<31)-1 {
		return (1 << 31) - 1
	}

	return resultInt
}

func run_myAtoi() {
	fmt.Println(myAtoi("1337c0d3"))
	fmt.Println(myAtoi("  -042"))
	fmt.Println(myAtoi("20000000000000000000"))
	fmt.Println(myAtoi("+1"))
}
